File: php/src/AssessService/WordScorer.php
<?php

declare(strict_types=1);

namespace FactChecker\AssessService;

class WordScorer implements Scorer
{
    const MAX_SCORE = 4;

    public function calculateScore(string $sentence): int
    {
        $score = $this->scoreWords($sentence);

        if ($score >= self::MAX_SCORE) {
            return self::MAX_SCORE;
        }

        return $score;
    }

    private function scoreWords(string $sentence): int
    {
        $count = preg_match_all('/\bcat(s)?\b/iSu', $sentence);

        return ($count === 0) ? 0 : $count + 1;
    }
}

File: java/app/src/main/java/factchecker/AssessService/WordScorer.java
package factchecker.AssessService;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WordScorer implements Scorer {
    public static final Integer MAX_SCORE = 4;

    public Integer calculateScore(String sentence) {
        Integer score = scoreWords(sentence);

        if (score >= MAX_SCORE) {
            return MAX_SCORE;
        }

        return score;
    }

    private Integer scoreWords(String sentence) {
        Pattern pattern = Pattern.compile("cats?");
        Matcher matcher = pattern.matcher(sentence);

        int count = Math.toIntExact(matcher.results().count());

        return (count == 0) ? 0 : count + 1;
    }
}
